### ISSUE TITLE
Keep double quotes for hyper links

### ISSUE THREAD

**Issue by tiffanyAtDuke** [tdewolff/minify#128]
I'm using v2.0. I noticed after applying minifying HTML, hyperlinks with embedded styles are in wrong formats. This is because that the double quotes around the link is removed, while the double quotes around other embedded styles are remained. 

For example, <a href="http://www.google.com" style="color:#ffffff; text-decoration:none;"> will be minified to <a href=http://www.google.com style="color:#ffffff; text-decoration:none;">, resulting in a broken link as http://www.google.comstyle=.

Can we keep double quotes around hyper links? Or at least can we have an extra option as "KeepDoubleQuotes"?

Attached here is the sample program:
[test.txt](https://github.com/tdewolff/minify/files/835409/test.txt)

**Comment by tdewolff** [comment]
`<a href="http://www.google.com" style="color:#ffffff; text-decoration:none;">Google</a>`

gets minified to

`<a href=http://www.google.com style=color:#fff;text-decoration:none>Google</a>`

But both are valid HTML as far as I know. The link still seems to work in Chrome. Is this the same output you are getting?

**Comment by tiffanyAtDuke** [comment]
Thanks for replying. The online minifier does provide the proper output: http://go.tacodewolff.nl/minify. But unfortunately this is not the same output I got from running my golang sample program (attached below). Could you please let me know how you get above result? Thanks!

I'm using golang with version 1.7.4 darwin/amd64. And I'm running the attached programs on my laptop: macOS Sierra Version 10.12.2.

I've tried both v1 and v2 syntax (html.Minify(m, "text/html", w, r) → html.Minify(m, w, r, nil)), and the minified result is always:

tiffanyfff$ go run test3.go
<!doctype html><title>Hyperlink Example</title><p>Click following link</p><a href=http://google.com style="color:#ffffff; text-decoration:none;">Google</a>

I've attached my sample program (both v1 and v2) here. Any comments/fixes/suggestions are welcome!
[test_v1.txt](https://github.com/tdewolff/minify/files/840266/test_v1.txt)
[test_v2.txt](https://github.com/tdewolff/minify/files/840267/test_v2.txt)

**Comment by tdewolff** [comment]
I reponded to your private e-mail. I assume you hadn't attached a CSS minifier and thus the quotes for the `style` attribute were still in place. However, either case is valid HTML. Please let me know what exact error you encounter in the browser, otherwise please close the issue.

**Comment by tiffanyAtDuke** [comment]
Thanks for replying. Yes, with the CSS minifier all double quotes will be removed and the minified HTML works fine in browser. 

However, some major email providers (email, outlook) cannot handle attributes without double-quote properly. So that 
hrefs=https://example.com/ style=color:#fff will becomes:
hrefs=https://example.com/+style=color:#fff

We'd be appreciated if there's support for an option to keep all the double quotes for attributes. So that all Email related HTML will be valid after minification. Thanks!

**Comment by mikegleasonjr** [comment]
Sorry to jump in but I guess there is little benefit to minify html that will be consumed only once by an email provider?  Even when you send a template to multiple recipients that's the problem of your email provider.

**Comment by tiffanyAtDuke** [comment]
Thanks for the comments. The initiates for us to apply minification is that some of the non-minified Emails cannot be properly displayed in browser, due to the large HTML size. 

With the KeepDoubleQuotes option, the usage of this library will be widely extended to HTML in Emails.

**Comment by mikegleasonjr** [comment]
@tiffanyAtDuke I see.

**Comment by tiffanyAtDuke** [comment]
I tested with all major email services on different platforms, and found out only fox mail can handle HTML without double quotes properly. All the other email services cannot manage minified Emails. 

A list of email service + platform that won't support attributes without double quotes:
Gmail (Chrome)
Gmail (Safari)
Gmail (iPhone App)
Outlook (Safari)
Outlook (Chrome)
AOL mail (Safari)
AOL mail (Chrome)

**Comment by tdewolff** [comment]
I have just tested this with Gmail but it works fine for me. I send an e-mail using telnet, with the following content:

`Content-Type: text/html; charset="ISO-8859-1"
<html>
<head>
<meta http-equiv='content-type' content='text/html; charset=ISO-8859-1'>
</head>
<body>
<p><a href=http://google.nl/ style="color:red">Test link</a></p>
</body>
</html>`

This is also what Gmail receives and shows in the original. This link works fine. However, I had problems with sending it to my own mail server because SpamAssassin blocks this apparently. I'm not sure if I do something wrong though, but it doesn't seem to like it. Therefore I couldn't test it in Outlook and Thunderbird.

However, Gmail seemed to work fine, so I'm really wondering why this doesn't seem to work for you...

**Comment by tdewolff** [comment]
@tiffanyAtDuke I wonder if this is really a problem of minify and the mail clients and not something else in your stack? As I cannot reproduce this I will close the issue, feel free to re-open if this still poses a problem.

**Comment by mkadin** [comment]
Closing the loop on this one - you're right that this did appear elsewhere in the stack - our downstream email sender. 

We reached out to them and they fixed the bug.  Apologies for the confusion here.

**Comment by chris-martin** [comment]
> We'd be appreciated if there's support for an option to keep all the double quotes for attributes.

I'd also really appreciate this; I'd like to use minify's other HTML shrinking features, but the quote-stripping seems excessing. The W3C HTML validator goes nuts at minify's HTML output. Some text editors' syntax highlighters get confused when there's a slash at the end of an unquoted attribute, as they think you're trying to close the tag.

**Comment by tdewolff** [comment]
I just tried the following minimal example in [W3C](https://validator.w3.org/nu/#textarea), and it doesn't give any warnings:

``` html
<!doctype html>
<title>.</title>
<a href=http://www.google.com/>Google</a>
```

This minifier follows HTML5 specs, see [Unquoted attribute value syntax](https://www.w3.org/TR/html5/syntax.html#unquoted) which specifies that this behavior is allowed. I think the problem is that your HTML source is not HTML5 but HTML4, possibly missing the DOCTYPE tag. It might be a wise decision for the minifier to throw a warning for non-HTML5 source code or to insert the doctype when it is missing. At this time, it replaces any doctype by one for HTML5, but this might be a little premature.

**Comment by brunoaugusto** [comment]
I understand this is an old issue by now, but I would like to reiterate the possibility of keeping quotes around hyperlinks because, as far as my knowledge goes, this seems to be preventing pages to be loaded by AJAX.

I manage a static website with a Search Tool that uses the Taxonomy Terms structure as data-source to find links matching some criteria which will then be loaded and parsed to build a UI. The structure of the Taxonomy Terms is something like that:

```
<ul>
    <li data-term="TaxonomyTermOfFirstLevel">TaxonomyTermOfFirstLevel
        <ul>
            <li data-term="TaxonomyTermOfSecondLevel">TaxonomyTermOfSecondLevel
                <ul>
                    <li data-term="TaxonomyTerm">
                        <a href="https://www.mydomain.com/some-page/">Some Page</a>
                    </li>
                </ul>
             </li>
         </ul>
      </li>
</ul>
```

The first Request that loads the page with the structure above works fine, but when trying to load any of the links, all Requests fail and raise the following error:

> Mixed Content: The page at 'https://www.mydomain.com/search/' was loaded over HTTPS, but requested an insecure XMLHttpRequest endpoint 'http://www.mydomain.com/some-page/'. This request has been blocked; the content must be served over HTTPS.

As you can see, both Requester (`/search`) and the Requested (`/some-page`) are in the same domain but the browser says the URL being requested is insecure because it's not under SSL. But if you look at the fragment I provided, the Requested URL **does** have the `https` schema prefix.

I've searched exhaustively about this without success and only when I deactivated the output minifying everything worked as intended. I then analyzed everything involving the searching process, from all of the HTML to each of the JS files responsible for the searching (after all it's a static site) and the only difference was the lack of quotes around the `href` attributes.

Thank you for your time

**Comment by tdewolff** [comment]
That seems very strange. Are you sure this is HTML5, ie. it has `<!DOCTYPE html>` at the top of the document? What happes if you use `//www.mydomain.com/some-page/` (drop the `http:` and `https:`)? That should automatically pick the same protocol as the main page.

If not, could you make a minimal working example that doesn't work as intended?

**Comment by brunoaugusto** [comment]
Very creepy, right?

Everything looks fine, DocType, markup, but I haven't considered dropping the scheme, I'll try that and see if it solves the issue. If not, I might keep this one-page without minifying, even though it's not the best course of action. Replicating the scenario manually — since the whole Taxonomy Page I mentioned is generated by Hugo — plus the Github Pages association to enforce the same domain condition would be kind of time-consuming.

Thank you for your time

**Comment by brunoaugusto** [comment]
Well, I tested removing the scheme directly from the HTML of the Template Hugo would use to create the Taxonomy Page and through the JavaScript as well (see below) and neither of the methods worked, without the quotes browser keeps blocking the Requests.

`let url = $( current ).find( 'a' ).attr( 'href' ).replace( /https?:/gi, '');`

Would you really, really need a full example?

**Comment by tdewolff** [comment]
Hm yes, sounds like a bug within the search tool you use...? Is the taxonomy terms page by any chance XML and not HTML? XML prohibits removing quotes around URLs.

Can you show me a (more complete) part of the taxonomy terms page and refer to the search tool that you use? I don't think Hugo has anything to do with it, as minification is performed afterwards, right? Can you show me the piece of JS that lookups the URL and makes the XMLHttpRequest?

**Comment by brunoaugusto** [comment]
Hugo is just the generator, there are no problems with it (this time). It outputs a very, very long HTML (not XML) as Taxonomy Page, but in the end, it's just several nested lists like the one I posted above:

```
<ul>
    <li data-term="TaxonomyTermOfFirstLevel">TaxonomyTermOfFirstLevel
        <ul>
            <li data-term="TaxonomyTermOfSecondLevel">TaxonomyTermOfSecondLevel
                <ul>
                    <li data-term="TaxonomyTerm">
                        <a href="https://www.mydomain.com/some-page/">Some Page</a>
                    </li>
                </ul>
             </li>
         </ul>
      </li>
</ul>
```

I make a Request for the Taxonomy Page, parse the response, iterate over the list and make N Requests for each link found:

```
load( '/taxonomy/' ).done( ( response ) => {

    let data = $( jQuery.parseHTML( response ) );

    $.each( data, ( offset, current ) => {

        let url = $( current ).find( 'a' ).attr( 'href' );

        load( url ).done( ( response ) => {
            // Do something with response data
        });
    });
});
```

For the reference, the `load` function is a very pseudo-cache for all XHR to not repeat the Requests in case more than one search is made:

```
let cache = {};

function load( url ) {

    if( ! cache[ url ] ) {
        cache[ url ] = $.ajax( url );
    }

    return this.cache[ url ];
}
```

With a non-minified HTML, it woks nicely. After minifiying, Chrome and Firefox block all Requests made withi the jQuery.each(). And the only difference between them are the spaces/new lines and the lack of quotes around the `href`

If that doesn't help I'll stop procrastinating and next Monday I'll build a full example :stuck_out_tongue:

**Comment by tdewolff** [comment]
I just tested something simple as:

```
<a href="https://example.com">One</a>
<a href=https://example.com>Two</a>

<script>
$('a').each(function (idx,c){
    url = $(c).attr('href');
    console.log(url);
});
</script>
```

but the `url` is the same in both cases. I don't think it is possible that it is related to the content of the `href` attribute. Perhaps some meta tag in the head section has changed? Perhaps the URL you're loaded over has changed? Please print the contents of the keys of `this.cache` to ensure that the URLs are in there in the right format.

**Comment by brunoaugusto** [comment]
Took me a little longer than I expected since I had to create meaningful fake content, but [here we go](https://github.com/brunoaugusto/playground).

In this repository, you'll find all files generated by Hugo for a very minimalistic site with a few models of Cars with their Brands and Country as Taxonomy. From [this URL](https://brunoaugusto.github.io/playground/search/), already active as a Github Pages, you can see a very simplified implementation of my real-case scenario.

The idea here is to load the [Taxonomy Terms Pages](https://brunoaugusto.github.io/playground/tags/), find all `<li>` tags, Request the URL of each link and simply appending the Requested Page Title in some element.

Since there's no way to enable Github Pages for two branches at once, the [current revision](https://github.com/brunoaugusto/playground/commit/0a4cc56598d7885c5d071fb346ca8832dd5a8ef1) is the one Hugo minified the output using your code/software (`hugo --minify` command) which means it **does not** work. Loading the Taxonomy Page works, but loading any of the URLs found, raises the Mixed Content error.

The [previous revision](https://github.com/brunoaugusto/playground/commit/90a57c1248e0dc395714f9da79367d1dcf38bdac), without minifying, works as intended. I'm not quite sure as I'm not very good with GIT, but I believe it will suffice for you to roll-back and see.

**Comment by mkadin** [comment]
Looked at this for a second, what appears to be happening, is your page is requesting `https://brunoaugusto.github.io/playground/cars/audi/a6` (note the missing trailing slash) which is then 301 redirecting to `https://brunoaugusto.github.io/playground/cars/audi/a6/`.  That redirect is done with http instead of https which is what is causing the problem.  

If you go to the page without the trailing slash, your browser will redirect you, but an XHR doesn't seem to have that behavior.

Here's a quick example from the chrome debugger in the console on your search page:

![image](https://user-images.githubusercontent.com/1679344/47732659-e0e6fa80-dc23-11e8-9b8a-be04f4aeebc7.png)

Questions / Suggestions:

1) Does minifying with this library do anything with trailing slashes?
2) Does not having the quotes at the end of urls do anything with trailing slashes?
3) You might want to make sure your code includes the slash, and things will probably just work.
4) This is a strange way to populate a webpage like this.  Suggest you host a JSON file with all of your car data instead of parsing another HTML page to get your data.

**Comment by brunoaugusto** [comment]
1. I just checked both commits and, minified or not, the URLs remains intact with the trailing slash as defined. I made sure to generate the HTML with them because I had this kind of trouble in past for not having the trailing slash in a different part of my Application.

2. I'm not sure what you mean, but I would say they don't interfere, after all, the just wrap the information.

3. They're all there in the HTML I create, the HTML Hugo generates and I don't add or remove anything in the JS. I use the `href` "as is". This is so true that "console-ing" (ugh) `url` after its definition outputs each one of them from the scheme and up to the trailing slash. Unfortunately, Hugo doesn't provide a way to run the minifier with the server altogether.

4. I wish but it isn't exactly easy to handle JSON with Hugo and if I choose to move everything to JS I would have to manually create a separate index of URLs or then change the Taxonomy Template into something less useful without JS, as it also acts as a secondary Sitemap. The real implementation allows me, for example, to query for multiple Taxonomy Terms (which I have more than just two) and thus intersect the results.

Right now while answering your questions I had the idea to trim ending slashes and manually append one in the JS and, as expected, it worked (latest commit):

`var url=$(current).find('a').attr('href').replace(/\/$/,"")+'/'`

But, as far as I can tell, I shouldn't have to do this on my own. It's like the trailing slash is being taken out of the equation at some point without reason.

**Comment by brunoaugusto** [comment]
All revisions always had trailing slashes, untouched.

Trimming and re-adding the slashes in the JS code as I showed above did the trick and I think it changes nothing only from our perspective. I can't explain the intricacies involved because I don't really know how browsers work in this matter, but if I had to guess I'd say the browsers are misinterpreting the trailing slash, perhaps considering it as part of the HTML structure instead of the attribute value and thus they're not "bringing it along" the rest of the attribute while following the redirection.

While having attributes unquoted is a possibility with HTML5 there are a few quirks one must take care of and, while I haven't seen this one anywhere else — at least not without being related a particular software — I believe it's another of these cases.
